### ISSUE TITLE
dataBreakpointInfo and setDatabreakpointsRequest needs new fields

### ISSUE THREAD

**Issue by theIDinside** [microsoft/debug-adapter-protocol#404]
dataBreakpointInfo request takes the following parameters:

```javascript
interface DataBreakpointInfoArguments {
  /**
   * Reference to the variable container if the data breakpoint is requested for
   * a child of the container. The `variablesReference` must have been obtained
   * in the current suspended state. See 'Lifetime of Object References' in the
   * Overview section for details.
   */
  variablesReference?: number;

  /**
   * The name of the variable's child to obtain data breakpoint information for.
   * If `variablesReference` isn't specified, this can be an expression.
   */
  name: string;

  /**
   * When `name` is an expression, evaluate it in the scope of this stack frame.
   * If not specified, the expression is evaluated in the global scope. When
   * `variablesReference` is specified, this property has no effect.
   */
  frameId?: number;
}
```

As we can see, frameId has no effect when variablesReference is passed in.

However, frameId will *never* have an effect - with or without variablesReference, because `setDatabreakpoints` can not differentiate between a `dataBreakpointInfo` that took either a `variablesReference` or a `frameId` parameter. Unless the DAP implementer make `dataBreakpointsInfo` be the actual request that *sets* the bp, these two requests make no sense (when talking specifically about `frameId`).

One solution would be to add another field to the response in `dataBreakpointInfo` that signals what "kind" of `dataBreakpointInfo` request & response was sent and received. That way, when it's time for `setDatabreakpoints` request to roll in, the DAP implementation can inspect "should this be set for this frame (because frameId was passed) or should it evaluate it as global?".

There is a solution to this, currently for DAP-implementers - and that's to customize (and thus break) the DAP specification when it comes to these two requests, by adding their own fields. But that's not ideal; particularly when the spec does not mention that these requests are runtime specific, which is the case for things like `attach` and `launch` requests respectively. The best approach would be if it was standardized in  the spec.

**Comment by connor4312** [comment]
> because setDatabreakpoints can not differentiate between a dataBreakpointInfo that took either a variablesReference or a frameId parameter

I don't think this is the case. `DataBreakpointInfoResponse` contains the adapter-defined `dataId` property, which is a string. The `DataBreakpoint` in `SetDataBreakpoints` then requires the `dataId`. It's up to the adapter to encode any necessary correlation they need into the data ID. For example, they might maintain a random map of UUIDs, or actually encode the frame ID along with other information as a JSON string inside the data ID.

**Comment by theIDinside** [comment]
> > because setDatabreakpoints can not differentiate between a dataBreakpointInfo that took either a variablesReference or a frameId parameter
> 
> I don't think this is the case. `DataBreakpointInfoResponse` contains the adapter-defined `dataId` property, which is a string. The `DataBreakpoint` in `SetDataBreakpoints` then requires the `dataId`. It's up to the adapter to encode any necessary correlation they need into the data ID. For example, they might maintain a random map of UUIDs, or actually encode the frame ID along with other information as a JSON string inside the data ID.

So instead of keeping a cohesive design across the requests that involve variable references, the design choice is to break with that convention?

It breaks the idea of a specification. If a debugger wants to implement the spec, now it has to introduce runtime specific parsing of `dataId`, that's beyond the spec, which introduces new complexity that also all DA-implementers for that debugger need to adhere to; amounting essentially to a subset of the spec.

I would certainly argue this is not ideal. Particularly since the DAP-spec does not discuss any runtime specific variability with respect to these requests. It's also a bit odd, that two out of all requests (that don't explicitly inform of runtime variability, like the `attach` and `launch` requests) introduce this variability without informing of it.

I would accept I guess a documentation update on this, though I would consider it the less good option of the two alternatives. The problem of variability, whether explicit or implicit, would be solved in its entirety with the added field.
